### ISSUE TITLE
Error in Linking Plugin with strapi-helper-plugin

### ISSUE THREAD

**Issue by diana4397** [strapi/strapi#2066]
I have generated plugin using  strapi generate:plugin my-plugin
=============================================================================== -->> debug.log=============================================================================== -->

**Informations**
- **Node.js version**: 8.9.4
- **npm version**: 5.8.0
- **Strapi version**: 3.0.0-alpha.14.2
- **Database**: Mongo
- **Operating system**: Ubuntu 

**What is the current behavior?**

I am getting error while linking my-plugin with strapi-helper-plugin.
gifsicle@3.0.4 postinstall /usr/lib/node_modules/strapi-helper-plugin/node_modules/gifsicle
> node lib/install.js

  ⚠ EACCES: permission denied, mkdir '/usr/lib/node_modules/strapi-helper-plugin/node_modules/gifsicle/vendor'
  ⚠ gifsicle pre-build test failed
  ℹ compiling from source
/usr/lib/node_modules/strapi-helper-plugin/node_modules/onetime/index.js:15
                                throw new Error(fnName + ' can only be called once.');
                                ^

Error: callback() can only be called once.
    at onetime (/usr/lib/node_modules/strapi-helper-plugin/node_modules/onetime/index.js:15:11)
    at /usr/lib/node_modules/strapi-helper-plugin/node_modules/download/index.js:156:5
    at ConcatStream.<anonymous> (/usr/lib/node_modules/strapi-helper-plugin/node_modules/concat-stream/index.js:37:43)
    at emitNone (events.js:111:20)
    at ConcatStream.emit (events.js:208:7)
    at finishMaybe (/usr/lib/node_modules/strapi-helper-plugin/node_modules/readable-stream/lib/_stream_writable.js:630:14)
    at endWritable (/usr/lib/node_modules/strapi-helper-plugin/node_modules/readable-stream/lib/_stream_writable.js:638:3)
    at ConcatStream.Writable.end (/usr/lib/node_modules/strapi-helper-plugin/node_modules/readable-stream/lib/_stream_writable.js:594:41)
    at DuplexWrapper.onend (/usr/lib/node_modules/strapi-helper-plugin/node_modules/readable-stream/lib/_stream_readable.js:577:10)
    at Object.onceWrapper (events.js:313:30)
npm WARN eslint-config-airbnb@12.0.0 requires a peer of eslint-plugin-import@^1.16.0 but none is installed. You must install peer dependencies yourself.
npm WARN eslint-config-airbnb-base@8.0.0 requires a peer of eslint-plugin-import@^1.16.0 but none is installed. You must install peer dependencies yourself.
npm WARN eslint-config-prettier@2.3.0 requires a peer of eslint@>=3.14.1 but none is installed. You must install peer dependencies yourself.
npm WARN eslint-plugin-redux-saga@0.1.5 requires a peer of redux-saga@>= 0.11.1 < 1 but none is installed. You must install peer dependenciesyourself.
npm WARN react-addons-test-utils@15.6.2 requires a peer of react-dom@^15.4.2 but none is installed. You must install peer dependencies yourself.
npm WARN scroll-behavior@0.7.0 requires a peer of history@^1.12.1 || ^2.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/strapi-helper-plugin/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! gifsicle@3.0.4 postinstall: `node lib/install.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the gifsicle@3.0.4 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.-

**Steps to reproduce the problem**

**What is the expected behavior?**

**Suggested solutions**

**Comment by lauriejim** [comment]
Can you update the body format of your issue to make it more readable please. It will help us to help you.

**Comment by soupette** [comment]
@diana4397  It also seems that you have permissions issues...

> ⚠ EACCES: permission denied, mkdir '/usr/lib/node_modules/strapi-helper-plugin/node_modules/gifsicle/vendor'

**Comment by diana4397** [comment]
@lauriejim 
Can you please explain in brief. I can't get it.

**Comment by maturanomx** [comment]
> @lauriejim
> Can you please explain in brief. I can't get it.

He is meaning about the format of your issue. Format your log as code, follow the template. Is hard to read as is now.

> @soupette
> I have used sudo also as there was a permission issue first but then also this error comes up every time.

... you should take care about the permissions system, for your own good, don't run `npm` (and almost nothing) as root.

Your issue is not with strapi but with the filesystem permission of your Unix SO. A dependency is trying compile something, this is running in a subshell child of a `npm run-script` and, [from the documentation](https://docs.npmjs.com/misc/scripts#user):

> If npm was invoked with root privileges, then it will change the uid to the user account or uid specified by the user config, which defaults to nobody. Set the unsafe-perm flag to run scripts with root privileges.

So, doesn't matter you run it as `root` because the user is changed. So, three options comes in my mind:

1. Do what the documentations said to be able to run the script as `root` (bad idea).

2. Set the permissions for `/usr/lib/node_modules/` to something that works for you.
  a) You can just add write permissions for everyone (... works but not the ideal)
  b) You can change the owner of that path to your user (... better is you are the only user)
  c) Use a group where your user is member and set the correct permissions (better, you understand you SO)

3. Manage your packages in your user space; the better of the options. You can configure `npm` or use a tool like [nvm](https://github.com/creationix/nvm) to don't have to deal with that.


### TLDR: Your problem is not strapi, the solution is from your SO

**Comment by soupette** [comment]
Thanks @maturanomx for your answer!
